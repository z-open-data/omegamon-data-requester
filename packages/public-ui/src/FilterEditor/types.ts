import type { Filter, FilterCombinationResult, ClauseFilter, WithTouched } from 'public-common';

export type ReplacementDefinition<CLAUSE> = [Filter<CLAUSE> | 'root', FilterCombinationResult<CLAUSE>];

export type ReplaceClauseFunction<CLAUSE> = (...replacements: Array<ReplacementDefinition<CLAUSE>>) => void;

export type OriginalAndEditedClause<CLAUSE> =
  | {
      /** Original clause user currently edits */
      originalClauseFilter: ClauseFilter<CLAUSE>;
      /** Current state of editing form(s). Once user applies the changes, it will be
       * converted to a MetricsQueryFilterClause and replace the original clause in
       * the whole filter
       */
      editedClause: CLAUSE;
    }
  | {
      originalClauseFilter: null;
      editedClause: null;
    };

export type BaseEditorAction<CLAUSE> =
  | {
      type: 'editing_started';
      clauseFilter: ClauseFilter<CLAUSE>;
    }
  | {
      type: 'editing_applied';
      clauseFilter: ClauseFilter<CLAUSE>;
    }
  | {
      type: 'clause_edited_partially';
      clauseFields: Partial<CLAUSE>;
    }
  | {
      type: 'clause_edited_fully';
      clause: CLAUSE;
    }
  | {
      type: 'clause_removed';
      clauseFilter: ClauseFilter<CLAUSE>;
    };

export type AddClauseAction<CLAUSE> =
  | {
      // For the case when there are no filters defined yet and user adds the first clause (root clause)
      type: 'new_clause_added';
      sourceFilter: null;
      connector: null;
      newClauseFilter: ClauseFilter<CLAUSE & WithTouched>;
    }
  | {
      /**
       * For the case when there is already at least one clause
       * and user clicks "Add new clause" button on a clause editor or a line connector.
       */
      type: 'new_clause_added';
      /**
       * If user clicked "Add new clause" button on:
       * - clause editor, then `sourceFilter` is that clause filter.
       * - line connector, then `sourceFilter` is that compound filter (filter containing property `and` / `or`).
       */
      sourceFilter: Filter<CLAUSE>;
      connector: 'or' | 'and';
      newClauseFilter: ClauseFilter<CLAUSE & WithTouched>;
    };

export type CancelClauseEditAction<CLAUSE> = {
  type: 'editing_canceled';
  clauseFilter: ClauseFilter<CLAUSE & WithTouched>;
};

export type EditorAction<CLAUSE> = BaseEditorAction<CLAUSE> | AddClauseAction<CLAUSE> | CancelClauseEditAction<CLAUSE>;

export type DispatchEdit<CLAUSE> = React.Dispatch<EditorAction<CLAUSE>>;
