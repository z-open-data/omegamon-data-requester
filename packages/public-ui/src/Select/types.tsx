/**
 * Defines a range of character position in a string.
 * Helps to track position for variable insert, highlight variables or errors in selected options.
 */
export type TextSelection = {
  startIdx: number;
  endIdx: number;
};

/**
 * apply function is adjustment on Option type to improve usage of it in string environment.
 * Based on function implementation, it will inject option value into string on selection position.
 * It collaborates a lot with OptionsPreprocessor.
 */
export type Option = {
  id: number | string;
  value: string;
  label?: string;
  disabled?: boolean;
  isNewCustom?: boolean;
  apply(this: Option, inputText: string, selection: TextSelection): string;
};

export type ComboboxState = {
  input?: string;
  options?: Option[];
};

export type OnChangePreprocessor = (options: Option[], input: string, selection: TextSelection) => ComboboxState;

export type OptionsGenerator = (
  options: Option[],
  selectedOptions: Option[],
  inputValue: string,
  cursorPosition?: number,
  allowCustomValues?: boolean
) => Option[];
