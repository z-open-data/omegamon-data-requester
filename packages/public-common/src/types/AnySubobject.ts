// (Recurcively) returns a union of any object-like field and T itself
//
// type A = {
//   a1: number;
//   a2: string;
// }
//
// type B = {
//   b1: A;
//   b2: string;
// }
//
// type C = {
//   c1: B;
//   c2: boolean;
// }
//
// AnySubobject<C> => A | B | C;

// KNOWN is used to break infinite looping over recursive structures
// It contains all the types we already have visited
type ExtractSubobjects<T, KNOWN> = T extends object ? (T extends KNOWN ? never : AnySubobject<T, KNOWN>) : never;
export type AnySubobject<T extends object, KNOWN = never> = T | ExtractSubobjects<T[keyof T], KNOWN | T>;
