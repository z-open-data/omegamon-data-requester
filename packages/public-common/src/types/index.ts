export type { AnySubobject } from './AnySubobject';
export type { Branded } from './Branded';
export type { DeepPick } from './DeepPick';
export type { DeepPropNames } from './DeepPropNames';
export type { DiscriminatedUnionMap } from './DiscriminatedUnionMap';
export type { FilterProperties, FilterPropertiesOut, PickNonNever } from './FilterProperties';
export type { NonEmptyArray } from './NonEmptyArray';
export type { OmitFromUnion } from './OmitFromUnion';
export type { SingleFieldOf } from './SingleFieldOf';
export type { UnionToIntersection } from './UnionToIntersection';
export type { OmitPropertyFromUnion } from './OmitPropertyFromUnion';
export type { RecursivePartial } from './RecursivePartial';
export type { Filter, ClauseFilter, CompositeFilter, AndFilter, OrFilter } from './Filter';
