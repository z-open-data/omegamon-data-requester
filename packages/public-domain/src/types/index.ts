export type { AffinityEntity, AffinityId } from './AffinityEntity';
export {
  MVS_CICS_AFFINITY_ID,
  SUPPORTED_AFFINITIES_FOR_METRICS,
  SUPPORTED_AFFINITIES_FOR_SITUATIONS,
  ITM_FRAMEWORK_AFFINITY_ID,
  ITM_FRAMEWORK_AFFINITY_ENTITY,
  TEMS_AFFINITY_ENTITY,
  TEMS_AFFINITY_ID,
  ZOS_APPLICATION_IDS,
  NETWORK_APPLICATION_IDS,
} from './AffinityEntity';
export {
  ITM_FRAMEWORK_APPLICATION_TABLES_FOR_SITUATIONS,
  ITM_FRAMEWORK_APPLICATION_FOR_SITUATIONS,
} from './ApplicationMetadata';
export type { ApplicationMetadata, ApplicationTableMetadata } from './ApplicationMetadata';
export type { ColumnMetadata, EnumMap } from './ColumnMetadata';
export { findEnumValue } from './ColumnMetadata';
export { TIME_UNITS } from './ColumnMetadata';
export { KNOWN_COLUMN_TYPES, WRITETIME_COLUMN_ID, writetimeColumnMetadata } from './ColumnMetadata';
export type {
  AvailableTableMetadataPropertyArray,
  ColumnMetadataMap,
  FilteredTableMetadata,
  TableMetadata,
  SampleType,
} from './TableMetadata';
export { validateTableMetadataFieldList } from './TableMetadata';
export type { Agent, Originnode, AgentIndex, GroupIndex, AgentsAndGroups, Group } from './AgentsAndGroups';
export { HUB_GROUP_NAME } from './AgentsAndGroups';
export type { ManagedSystems } from './ManagedSystems';
export * from './Action';
