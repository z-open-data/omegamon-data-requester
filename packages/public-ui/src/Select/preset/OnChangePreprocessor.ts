// eslint-disable-next-line no-restricted-imports
import { isInsideVariable } from '../VariableUtils';

import { OnChangePreprocessor, Option, TextSelection } from '~/Select';

export const onChangePreprocessor: OnChangePreprocessor = (
  options: Option[],
  input: string,
  selection: TextSelection
) => {
  const newOption = [...options].pop();
  const variableState = isInsideVariable(input, selection.startIdx);

  if (variableState.insideVariable && newOption) {
    return {
      input: newOption.apply(input, { startIdx: variableState.varStart, endIdx: variableState.varEnd }),
    };
  }
  return {
    options: options,
    input: '',
  };
};
