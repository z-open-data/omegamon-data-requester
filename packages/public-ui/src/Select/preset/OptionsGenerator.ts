// eslint-disable-next-line no-restricted-imports
import { isInsideVariable, extractVarValue } from '../VariableUtils';

import { OptionsGenerator, Option } from '~/Select';

export const generateOptions: OptionsGenerator = (
  options: Option[],
  selectedOptions: Option[],
  input: string,
  cursorPosition?: number,
  allowCustomValues?: boolean
) => {
  const variableState = isInsideVariable(input, cursorPosition || 0);
  if (variableState.insideVariable) {
    const stringToFind = extractVarValue(input, variableState).toLowerCase();

    return options.filter(
      (option) => option.value.startsWith('${') && option.value.toLowerCase().includes(stringToFind)
    );
  }

  const filteredItems = options.filter((item) => {
    return (
      (item.label?.toLowerCase().includes(input.toLowerCase()) ||
        item.value.toLowerCase().includes(input.toLowerCase())) &&
      !selectedOptions?.includes(item)
    );
  });

  if (
    input.length > 0 &&
    allowCustomValues &&
    !selectedOptions?.some((item) => {
      return item.value === input;
    })
  ) {
    filteredItems.push({
      id: input,
      value: input,
      label: input,
      isNewCustom: true,
      apply: (i, _) => i,
    });
  }
  return filteredItems;
};
