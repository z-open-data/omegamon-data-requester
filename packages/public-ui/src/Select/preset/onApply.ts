import { TextSelection, Option } from '~/Select';

export function sharedApply(this: Option, input: string, selection: TextSelection) {
  const value =
    input.substring(0, selection.startIdx) +
    this.value +
    (selection.endIdx > 0 ? input.substring(selection.endIdx + 1) : '');
  return value;
}
