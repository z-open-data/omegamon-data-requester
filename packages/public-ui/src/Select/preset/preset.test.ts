/* eslint-disable no-restricted-imports */
import { Option } from '../types';

import { sharedApply } from './onApply';
import { onChangePreprocessor } from './OnChangePreprocessor';
import { generateOptions } from './OptionsGenerator';

export const options: Option[] = [
  {
    label: 'DB2 Subsystems',
    value: 'KDP',
    id: '%IBM.STATIC017',
    apply: sharedApply,
  },
  {
    label: 'MVS CICSTG',
    value: 'KGW',
    id: '%IBM.STATIC115',
    apply: sharedApply,
  },
  {
    label: 'IMS Subsystems',
    value: 'KIP',
    id: '%IBM.STATIC014',
    apply: sharedApply,
  },
];

export const variables: Option[] = [
  {
    id: 'db2id',
    value: '${db2id}',
    label: 'DB2 variable',
    apply: sharedApply,
  },
];

describe('OnChangePreprocessor multi', () => {
  it('empty state', () => {
    expect(onChangePreprocessor([], '', { startIdx: 0, endIdx: 0 })).toMatchSnapshot();
  });
  it('empty input', () => {
    expect(onChangePreprocessor(options, '', { startIdx: 0, endIdx: 0 })).toMatchSnapshot();
  });
  it('input is not empty and in variables', () => {
    expect(
      onChangePreprocessor([...options, ...variables], 'testOfVar:${', { startIdx: 12, endIdx: 12 })
    ).toMatchSnapshot();
  });
  it('input is not empty and in closed variables', () => {
    expect(
      onChangePreprocessor([...options, ...variables], 'testOfVar:${}', { startIdx: 12, endIdx: 12 })
    ).toMatchSnapshot();
  });
  it('input is not empty and not in variables', () => {
    expect(
      onChangePreprocessor([...options, ...variables], 'testOfVar:${}', { startIdx: 13, endIdx: 13 })
    ).toMatchSnapshot();
  });
});

describe('Options Generator', () => {
  it('empty state', () => {
    expect(generateOptions([], [], '')).toMatchSnapshot();
  });
  it('empty input', () => {
    expect(generateOptions(options, [], '')).toMatchSnapshot();
  });
  it('empty input, few options selected, all selected', () => {
    expect(generateOptions(options, [options[0]], '')).toMatchSnapshot();
  });
  it('empty input, all selected', () => {
    expect(generateOptions(options, options, '')).toMatchSnapshot();
  });
  it('some input, all selected, custom values are not allowed', () => {
    expect(generateOptions(options, options, 'testOfFunctions')).toMatchSnapshot();
  });
  it('some input, all selected, custom values are allowed', () => {
    const testInput = 'testOfFunctions';
    expect(generateOptions(options, options, testInput, 0, true)).toMatchSnapshot();
  });

  it('some input, some selected, no options returns', () => {
    expect(generateOptions(options, [options[0]], 'KD')).toMatchSnapshot();
  });

  it('some input, some selected, one option returns', () => {
    expect(generateOptions(options, [options[1]], 'KD')).toMatchSnapshot();
  });

  it('some input, some selected, few options returns', () => {
    expect(generateOptions(options, [options[1]], 'K')).toMatchSnapshot();
  });

  it('some input, some selected, allows custom - custom option returns', () => {
    const data = generateOptions(options, [options[0]], 'KD', 0, true);
    expect(data).toMatchSnapshot();
    expect(data[0].isNewCustom).toBe(true);
  });

  it('some input, some selected, allows custom - one and custom options returns', () => {
    const data = generateOptions(options, [options[1]], 'KD', 0, true);
    const customElement = data.find((el) => el.isNewCustom);
    expect(data).toMatchSnapshot();
    expect(customElement).toMatchSnapshot();
  });

  it('some input, some selected, allows custom - few and custom options returns', () => {
    const data = generateOptions(options, [options[1]], 'K', 0, true);
    expect(data).toMatchSnapshot();
  });
});
