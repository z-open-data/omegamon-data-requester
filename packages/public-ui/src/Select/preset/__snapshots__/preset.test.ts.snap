// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnChangePreprocessor multi empty input 1`] = `
{
  "input": "",
  "options": [
    {
      "apply": [Function],
      "id": "%IBM.STATIC017",
      "label": "DB2 Subsystems",
      "value": "KDP",
    },
    {
      "apply": [Function],
      "id": "%IBM.STATIC115",
      "label": "MVS CICSTG",
      "value": "KGW",
    },
    {
      "apply": [Function],
      "id": "%IBM.STATIC014",
      "label": "IMS Subsystems",
      "value": "KIP",
    },
  ],
}
`;

exports[`OnChangePreprocessor multi empty state 1`] = `
{
  "input": "",
  "options": [],
}
`;

exports[`OnChangePreprocessor multi input is not empty and in closed variables 1`] = `
{
  "input": "testOfVar:\${db2id}",
}
`;

exports[`OnChangePreprocessor multi input is not empty and in variables 1`] = `
{
  "input": "testOfVar:\${db2id}",
}
`;

exports[`OnChangePreprocessor multi input is not empty and not in variables 1`] = `
{
  "input": "",
  "options": [
    {
      "apply": [Function],
      "id": "%IBM.STATIC017",
      "label": "DB2 Subsystems",
      "value": "KDP",
    },
    {
      "apply": [Function],
      "id": "%IBM.STATIC115",
      "label": "MVS CICSTG",
      "value": "KGW",
    },
    {
      "apply": [Function],
      "id": "%IBM.STATIC014",
      "label": "IMS Subsystems",
      "value": "KIP",
    },
    {
      "apply": [Function],
      "id": "db2id",
      "label": "DB2 variable",
      "value": "\${db2id}",
    },
  ],
}
`;

exports[`Options Generator empty input 1`] = `
[
  {
    "apply": [Function],
    "id": "%IBM.STATIC017",
    "label": "DB2 Subsystems",
    "value": "KDP",
  },
  {
    "apply": [Function],
    "id": "%IBM.STATIC115",
    "label": "MVS CICSTG",
    "value": "KGW",
  },
  {
    "apply": [Function],
    "id": "%IBM.STATIC014",
    "label": "IMS Subsystems",
    "value": "KIP",
  },
]
`;

exports[`Options Generator empty input, all selected 1`] = `[]`;

exports[`Options Generator empty input, few options selected, all selected 1`] = `
[
  {
    "apply": [Function],
    "id": "%IBM.STATIC115",
    "label": "MVS CICSTG",
    "value": "KGW",
  },
  {
    "apply": [Function],
    "id": "%IBM.STATIC014",
    "label": "IMS Subsystems",
    "value": "KIP",
  },
]
`;

exports[`Options Generator empty state 1`] = `[]`;

exports[`Options Generator some input, all selected, custom values are allowed 1`] = `
[
  {
    "apply": [Function],
    "id": "testOfFunctions",
    "isNewCustom": true,
    "label": "testOfFunctions",
    "value": "testOfFunctions",
  },
]
`;

exports[`Options Generator some input, all selected, custom values are not allowed 1`] = `[]`;

exports[`Options Generator some input, some selected, allows custom - custom option returns 1`] = `
[
  {
    "apply": [Function],
    "id": "KD",
    "isNewCustom": true,
    "label": "KD",
    "value": "KD",
  },
]
`;

exports[`Options Generator some input, some selected, allows custom - few and custom options returns 1`] = `
[
  {
    "apply": [Function],
    "id": "%IBM.STATIC017",
    "label": "DB2 Subsystems",
    "value": "KDP",
  },
  {
    "apply": [Function],
    "id": "%IBM.STATIC014",
    "label": "IMS Subsystems",
    "value": "KIP",
  },
  {
    "apply": [Function],
    "id": "K",
    "isNewCustom": true,
    "label": "K",
    "value": "K",
  },
]
`;

exports[`Options Generator some input, some selected, allows custom - one and custom options returns 1`] = `
[
  {
    "apply": [Function],
    "id": "%IBM.STATIC017",
    "label": "DB2 Subsystems",
    "value": "KDP",
  },
  {
    "apply": [Function],
    "id": "KD",
    "isNewCustom": true,
    "label": "KD",
    "value": "KD",
  },
]
`;

exports[`Options Generator some input, some selected, allows custom - one and custom options returns 2`] = `
{
  "apply": [Function],
  "id": "KD",
  "isNewCustom": true,
  "label": "KD",
  "value": "KD",
}
`;

exports[`Options Generator some input, some selected, few options returns 1`] = `
[
  {
    "apply": [Function],
    "id": "%IBM.STATIC017",
    "label": "DB2 Subsystems",
    "value": "KDP",
  },
  {
    "apply": [Function],
    "id": "%IBM.STATIC014",
    "label": "IMS Subsystems",
    "value": "KIP",
  },
]
`;

exports[`Options Generator some input, some selected, no options returns 1`] = `[]`;

exports[`Options Generator some input, some selected, one option returns 1`] = `
[
  {
    "apply": [Function],
    "id": "%IBM.STATIC017",
    "label": "DB2 Subsystems",
    "value": "KDP",
  },
]
`;
